src/PackageManagement/NuGetPackageManager.cs

```diff
@@ -1638,27 +1638,62 @@ private Task OpenReadmeFile(INuGetProjectContext nuGetProjectContext, Cancellati
             }
 
             token.ThrowIfCancellationRequested();
-            nuGetProjectContext.Log(ProjectManagement.MessageLevel.Info, string.Format(Strings.RestoringPackage, packageIdentity));
+
+            nuGetProjectContext.Log(
+                ProjectManagement.MessageLevel.Info,
+                string.Format(Strings.RestoringPackage, packageIdentity));
+
             var enabledSources = (sourceRepositories != null && sourceRepositories.Any()) ? sourceRepositories :
                 SourceRepositoryProvider.GetRepositories().Where(e => e.PackageSource.IsEnabled);
 
             token.ThrowIfCancellationRequested();
 
-            using (var downloadResult = await PackageDownloader.GetDownloadResourceResultAsync(enabledSources,
-                packageIdentity,
-                Settings,
-                token))
+            // maximum number of times to try restoring the package
+            var maxTries = 3;
+            var success = false;
+
+            for (int i = 0; i < maxTries && !token.IsCancellationRequested && !success; i++)
             {
-                packageIdentity = downloadResult.PackageReader.GetIdentity();
+                try
+                {
+                    using (var downloadResult = await PackageDownloader.GetDownloadResourceResultAsync(
+                        enabledSources,
+                        packageIdentity,
+                        Settings,
+                        token))
+                    {
+                        packageIdentity = downloadResult.PackageReader.GetIdentity();
+
+                        // If you already downloaded the package, just restore it, don't cancel the operation now
+                        await PackagesFolderNuGetProject.InstallPackageAsync(
+                            packageIdentity,
+                            downloadResult,
+                            nuGetProjectContext,
+                            token);
 
-                // If you already downloaded the package, just restore it, don't cancel the operation now
-                await PackagesFolderNuGetProject.InstallPackageAsync(packageIdentity, downloadResult, nuGetProjectContext, token);
+                        success = true;
+                    }
+                }
+                catch (Exception) when (i < (maxTries - 1))
```

Comment:
Do we want to retry on all exception? it could be that we are retying on something that we can't recover from can we catch a better exception type? like web exception or http exception?